<a href="https://colab.research.google.com/github/jennifernnorris-ux/Jennifer_INFO4670_Spring2026/blob/main/Assignment_2_1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#1. Data Quality Check

**1.1 Using Python (pandas, matplotlib, or seaborn), load and inspect the Assignment 2 dataset.**

In [2]:
from google.colab import files
uploaded = files.upload()


Saving Assignment 2 dataset.csv to Assignment 2 dataset.csv


In [3]:
import pandas as pd

In [4]:
df = pd.read_csv("Assignment 2 dataset.csv")

In [5]:
df.head()

,Column 1,Date,AveragePrice,Total Volume,4046,4225,4770,Total Bags,Small Bags,Large Bags,XLarge Bags,type,year,region
0,0,12-27-2015,1.33,64236.62,1036.74,54454.85,48.16,8696.87,8603.62,93.25,0.0,conventional,2015,Albany
1,1,12-20-2015,1.35,54876.98,674.28,44638.81,58.33,9505.56,9408.07,97.49,0.0,conventional,2015,Albany
2,2,12-13-2015,0.93,118220.22,794.70,109149.67,130.50,8145.35,8042.21,103.14,0.0,conventional,2015,Albany
3,3,12-6-2015,1.08,78992.15,1132.00,71976.41,72.58,5811.16,5677.40,133.76,0.0,conventional,2015,Albany
4,4,11-29-2015,1.28,51039.60,941.48,43838.39,75.78,6183.95,5986.26,197.69,0.0,conventional,2015,Albany


In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 18254 entries, 0 to 18253
Data columns (total 14 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Column 1      18254 non-null  int64  
 1   Date          18254 non-null  object 
 2   AveragePrice  18254 non-null  float64
 3   Total Volume  18253 non-null  float64
 4   4046          18252 non-null  float64
 5   4225          18253 non-null  float64
 6   4770          18253 non-null  float64
 7   Total Bags    18253 non-null  float64
 8   Small Bags    18252 non-null  float64
 9   Large Bags    18252 non-null  float64
 10  XLarge Bags   18253 non-null  float64
 11  type          18253 non-null  object 
 12  year          18254 non-null  int64  
 13  region        18254 non-null  object 
dtypes: float64(9), int64(2), object(3)
memory usage: 1.9+ MB


**Write code to explore the data distribution (e.g., region, type, year) and check whether there is any bias. Provide both the code and your interpretation.**

In [7]:
df["region"].value_counts()

,count
region,
WestTexNewMexico,340
Albany,338
BaltimoreWashington,338
Boise,338
Boston,338
Atlanta,338
California,338
Charlotte,338
Chicago,338


When searching for bias in the regions, I found that West Tex New Mexico has 340 entries while the rest has 338. This is an outlier in the dataset, but does not present a strong region bias, only slight.

In [8]:
df["type"].value_counts()

,count
type,
organic,9127
conventional,9126


I checked the distribution of avocado types in the dataset. The number of organic and conventional avocados are similar, which indicates little to no bias in avocado type representation.

In [9]:
df["year"].value_counts()

,count
year,
2017,5722
2016,5616
2015,5615
2018,1300
1904,1


I examined the distribution of data across years. The years 2017, 2016, and 2015 contain more records than others, which indicates a temporal bias in the dataset. The data is not evenly distributed across all years.

**1.2 Write Python code to check for duplicate rows and missing values in the dataset. Show the number of duplicates and missing values for each column. Then, explain (in comments or markdown) how you would handle these issues (e.g., drop, impute, or replace).**

In [10]:
df[df.duplicated()]


,Column 1,Date,AveragePrice,Total Volume,4046,4225,4770,Total Bags,Small Bags,Large Bags,XLarge Bags,type,year,region
18249,10,1-14-2018,1.93,16205.22,1527.63,2981.04,727.01,10969.54,10919.54,50.00,0.0,organic,2018,WestTexNewMexico
18253,7,2-4-2018,1.63,17074.83,2046.96,1529.20,0.00,13498.67,13066.82,431.85,0.0,organic,2018,WestTexNewMexico


This dataset contains 2 duplicate rows. Duplicate records should be dropped to prevent skewing the analysis.

In [11]:
df.isnull().sum()

,0
Column 1,0
Date,0
AveragePrice,0
Total Volume,1
4046,2
4225,1
4770,1
Total Bags,1
Small Bags,2
Large Bags,2


This dataset contains a few instances of missing data. Numerical data such as 4046, 4225, 4770, Total volume, small bags, large bags, and XLarge Bags should be imputed with their median value to preserve the dataset size while minimizing distortion. Categorical values such as type should be replaced with a placeholder or imputed with the mode.

**1.3 Use Python code to print the number of rows and columns in the dataset (e.g., with df.shape). Based on the dataset size, explain (briefly) whether you think the dataset is sufficient for training a machine learning model.**

In [12]:
df.shape



(18254, 14)

The dataset contains over 18,000 rows and multiple relevant features, which provides a substantial amount of data for training a machine learning model. This size is generally sufficient to identify patterns and relationships in the data without severe risk of overfitting.

#2. Data Cleaning and Preprocessing

**2.1 Remove the first column or “Column 1” from the dataset. Treat the ‘year’ variable as nominal.**

In [13]:
df = df.drop(columns=["Column 1"])

In [14]:
df["year"] = df["year"].astype("category")

**2.2 Check for duplicate values and remove them.**

In [15]:
df.duplicated().sum()

np.int64(2)

In [16]:
df = df.drop_duplicates()

In [17]:
df.duplicated().sum()

np.int64(0)

**2.3 Check for missing values. If a data record (row) only has a few missing values, replace the missing values with the median of the column feature in that specific “Region” variable. If most column values in a data record are missing, remove the data record.**

In [18]:
df = df.copy()

In [19]:
df.isnull().sum()

,0
Date,0
AveragePrice,0
Total Volume,1
4046,2
4225,1
4770,1
Total Bags,1
Small Bags,2
Large Bags,2
XLarge Bags,1


In [20]:
bag_columns = ["Total Volume", "4046", "4225", "4770", "Total Bags", "Small Bags", "Large Bags", "XLarge Bags"]

for col in bag_columns:
    df[col] = df[col].fillna(
        df.groupby("region")[col].transform("median")
    )

In [21]:
df["type"] = df["type"].fillna(
    df.groupby("region")["type"].transform(lambda x: x.mode()[0])
)

In [22]:
df[["Total Volume", "4046", "4225", "4770", "Total Bags", "Small Bags", "Large Bags", "XLarge Bags", "type"]].isnull().sum()


,0
Total Volume,0
4046,0
4225,0
4770,0
Total Bags,0
Small Bags,0
Large Bags,0
XLarge Bags,0
type,0


**2.4 Find the correlation between the variables and describe how the correlated values among the variables impact the model accuracy.**


In [23]:
numeric_df = df.select_dtypes(include=["number"])

In [24]:
correlation_matrix = numeric_df.corr()
correlation_matrix

,AveragePrice,Total Volume,4046,4225,4770,Total Bags,Small Bags,Large Bags,XLarge Bags
AveragePrice,1.000000,-0.192770,-0.208327,-0.172948,-0.179462,-0.177106,-0.174745,-0.172958,-0.117608
Total Volume,-0.192770,1.000000,0.977863,0.974181,0.872203,0.963047,0.967238,0.880640,0.747159
4046,-0.208327,0.977863,1.000000,0.926110,0.833390,0.920057,0.925280,0.838646,0.699378
4225,-0.172948,0.974181,0.926110,1.000000,0.887856,0.905788,0.916032,0.810017,0.688811
4770,-0.179462,0.872203,0.833390,0.887856,1.000000,0.792316,0.802734,0.698473,0.679863
Total Bags,-0.177106,0.963047,0.920057,0.905788,0.792316,1.000000,0.994334,0.943009,0.804234
Small Bags,-0.174745,0.967238,0.925280,0.916032,0.802734,0.994334,1.000000,0.902589,0.806846
Large Bags,-0.172958,0.880640,0.838646,0.810017,0.698473,0.943009,0.902589,1.000000,0.710860
XLarge Bags,-0.117608,0.747159,0.699378,0.688811,0.679863,0.804234,0.806846,0.710860,1.000000


Strongly correlated values are those that are above 0.7 or below -0.7. Several numeric features show strong positive correlations, particularly among bag-related variables and total volume. These correlations suggest redundancy in the data, which can reduce model accuracy by increasing overfitting. Reducing or combining correlated features can improve model performance.

#3. Exploratory Data Analysis (EDA)


**3.1 Describe the variables**
- **Describe all variables in the dataset.**
- **For continuous variables: report *range (min, max), mean, median, and distribution.***
- **For categorical variables: list unique values.**

In [25]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 18252 entries, 0 to 18252
Data columns (total 13 columns):
 #   Column        Non-Null Count  Dtype   
---  ------        --------------  -----   
 0   Date          18252 non-null  object  
 1   AveragePrice  18252 non-null  float64 
 2   Total Volume  18252 non-null  float64 
 3   4046          18252 non-null  float64 
 4   4225          18252 non-null  float64 
 5   4770          18252 non-null  float64 
 6   Total Bags    18252 non-null  float64 
 7   Small Bags    18252 non-null  float64 
 8   Large Bags    18252 non-null  float64 
 9   XLarge Bags   18252 non-null  float64 
 10  type          18252 non-null  object  
 11  year          18252 non-null  category
 12  region        18252 non-null  object  
dtypes: category(1), float64(9), object(3)
memory usage: 1.8+ MB


In [26]:
numeric_cols = df.select_dtypes(include=["int64", "float64"]).columns
categorical_cols = df.select_dtypes(include=["object"]).columns

numeric_cols, categorical_cols

(Index(['AveragePrice', 'Total Volume', '4046', '4225', '4770', 'Total Bags',
        'Small Bags', 'Large Bags', 'XLarge Bags'],
       dtype='object'),
 Index(['Date', 'type', 'region'], dtype='object'))

In [27]:
df[numeric_cols].describe()

,AveragePrice,Total Volume,4046,4225,4770,Total Bags,Small Bags,Large Bags,XLarge Bags
count,18252.000000,1.825200e+04,1.825200e+04,1.825200e+04,1.825200e+04,1.825200e+04,1.825200e+04,1.825200e+04,18252.000000
mean,1.406032,8.505220e+05,2.929755e+05,2.951086e+05,2.283613e+04,2.396045e+05,1.821716e+05,5.432955e+04,3105.915918
std,0.402668,3.453275e+06,1.264888e+06,1.204027e+06,1.074556e+05,9.861651e+05,7.461194e+05,2.439468e+05,17691.485283
min,0.440000,8.456000e+01,0.000000e+00,0.000000e+00,0.000000e+00,0.000000e+00,0.000000e+00,0.000000e+00,0.000000
25%,1.100000,1.084172e+04,8.544900e+02,3.008553e+03,0.000000e+00,5.089967e+03,2.852128e+03,1.274625e+02,0.000000
50%,1.370000,1.073655e+05,8.646205e+03,2.905888e+04,1.849950e+02,3.974118e+04,2.636942e+04,2.648225e+03,0.000000
75%,1.660000,4.329334e+05,1.110664e+05,1.500853e+05,6.239325e+03,1.107766e+05,8.333329e+04,2.199632e+04,132.297500
max,3.250000,6.250565e+07,2.274362e+07,2.047057e+07,2.546439e+06,1.937313e+07,1.338459e+07,5.719097e+06,551693.650000


In [28]:
for col in categorical_cols:
    print(f"{col}:")
    print(df[col].unique())
    print()


Date:
['12-27-2015' '12-20-2015' '12-13-2015' '12-6-2015' '11-29-2015'
 '11-22-2015' '11-15-2015' '11-8-2015' '11-1-2015' '10-25-2015'
 '10-18-2015' '10-11-2015' '10-4-2015' '9-27-2015' '9-20-2015' '9-13-2015'
 '9-6-2015' '8-30-2015' '8-23-2015' '8-16-2015' '8-9-2015' '8-2-2015'
 '7-26-2015' '7-19-2015' '7-12-2015' '7-5-2015' '6-28-2015' '6-21-2015'
 '6-14-2015' '6-7-2015' '5-31-2015' '5-24-2015' '5-17-2015' '5-10-2015'
 '5-3-2015' '4-26-2015' '4-19-2015' '4-12-2015' '4-5-2015' '3-29-2015'
 '3-22-2015' '3-15-2015' '3-8-2015' '3-1-2015' '2-22-2015' '2-15-2015'
 '2-8-2015' '2-1-2015' '1-25-2015' '1-18-2015' '1-11-2015' '1-4-2015'
 '12-25-2016' '12-18-2016' '12-11-2016' '12-4-2016' '11-27-2016'
 '11-20-2016' '11-13-2016' '11-6-2016' '10-30-2016' '10-23-2016'
 '10-16-2016' '10-9-2016' '10-2-2016' '9-25-2016' '9-18-2016' '9-11-2016'
 '9-4-2016' '8-28-2016' '8-21-2016' '8-14-2016' '8-7-2016' '7-31-2016'
 '7-24-2016' '7-17-2016' '7-10-2016' '7-3-2016' '6-26-2016' '6-19-2016'
 '6-12-2016' '6-5

The type variable has two categories: conventional and organic. The region variable contains multiple geographic regions.

**3.2 Inspect the earliest recorded date**

-  **Find the earliest** `Date`.
- **Check if there are avocado prices recorded from the earliest date up to 2010.**
- **Comment: does the earliest data point look reasonable? Keep or remove?**

In [29]:
df["Date"] = pd.to_datetime(df["Date"])

In [30]:
df["Date"].min()

Timestamp('1904-01-21 00:00:00')

In [31]:
df[df["Date"].dt.year <= 2010].shape



(1, 13)

The earliest data point doesn't seem as reasonable because there is only one value before 2010, it should probably be removed

**3.3 Highest average price**
- **Find the highest value in "AveragePrice".**
- **Report which region it belongs to.**
- **Describe how you obtained the result.**

In [32]:
df["AveragePrice"].max()

3.25

In [33]:
df.loc[df["AveragePrice"].idxmax(), ["region", "AveragePrice"]]


,14125
region,SanFrancisco
AveragePrice,3.25


I obtained the highest average price by using the .max() function on the AveragePrice column and then filtering the dataset to identify the corresponding region.

**3.4 Highest total volume**
- **Find the highest total volume of avocados.**
- **Report which region it belongs to.**
- **Describe how you obtained the result.**

In [34]:
df["Total Volume"].max()


62505646.52

In [35]:
df.loc[df["Total Volume"].idxmax(), ["region", "Total Volume"]]

,9097
region,TotalUS
Total Volume,62505646.52


The highest total avocado volume was identified by calculating the maximum value in the Total Volume column and locating the associated region.